### PR TITLE
Adding SpyBOAT tool

### PR DESCRIPTION
**Pull request by tensionhead**
FOR CONTRIBUTOR:
* [x] - I have read the [CONTRIBUTING.md](https://github.com/galaxyproject/tools-iuc/blob/master/CONTRIBUTING.md) document and this tool is appropriate for the tools-iuc repo.
* [x] - License permits unrestricted use (educational + commercial)
* [x] - This PR adds a new tool or tool collection
* [ ] - This PR updates an existing tool or tool collection
* [ ] - This PR does something else (explain below)

### CODE REVIEW

File: tools/spyboat/SpyBOAT.xml [hunk 1]
```diff
@@ -0,0 +1,248 @@
+<tool id="spyboat" name="SpyBOAT" version="@TOOL_VERSION@">
+<description>wavelet transforms image stacks</description>
+  <macros>
+    <token name="@TOOL_VERSION@">0.1.1</token>
+  </macros>
+  
+    <requirements>
+      <requirement type="package" version="@TOOL_VERSION@">spyboat</requirement>
+    </requirements>
+    <version_command>python $__tool_directory__/spyboat_cli.py --version</version_command>
```
Comment (bgruening) [inline tools/spyboat/SpyBOAT.xml:10]: All paths and data-parameters should be single-quoted for security reasons.

File: tools/spyboat/SpyBOAT.xml [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+    <command detect_errors="exit_code"><![CDATA[
+        python $__tool_directory__/spyboat_cli.py
```
Comment (bgruening) [inline tools/spyboat/SpyBOAT.xml:12]: single-quotes

File: tools/spyboat/SpyBOAT.xml [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+          --input_path '$movie'
+          #if $gauss_sigma:
+             --gauss_sigma $gauss_sigma
+          #end if
+          #if $rescale_factor:
+              --rescale $rescale_factor
+          #end if
+
+          #if $masking.selection_mode == "create_static_mask":
```
Comment (bgruening) [inline tools/spyboat/SpyBOAT.xml:21]: the indentation is off here ...

Comment (bgruening) [reply tools/spyboat/SpyBOAT.xml:21]: please use 4spaces instead of one tab

File: tools/spyboat/SpyBOAT.xml [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+          --amplitude_out $amplitude_out
+	  #if $preprocessed_out:
+	     --preprocessed_out $preprocessed_out
+	  #end if
+
+	  --html_fname $html_out
+	  --report_img_path '$html_out.extra_files_path'
+
+          --ncpu "\${GALAXY_SLOTS:-8}"
+          > $log
```
Comment (bgruening) [inline tools/spyboat/SpyBOAT.xml:53]: single-quote

File: tools/spyboat/SpyBOAT.xml [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+
+          --phase_out $phase_out
+          --period_out $period_out
+          --power_out $power_out
+          --amplitude_out $amplitude_out
+	  #if $preprocessed_out:
+	     --preprocessed_out $preprocessed_out
+	  #end if
+
+	  --html_fname $html_out
```
Comment (bgruening) [inline tools/spyboat/SpyBOAT.xml:49]: single-quote

File: tools/spyboat/SpyBOAT.xml (hunk 1 above)
Comment (bernt-matthias) [inline tools/spyboat/SpyBOAT.xml:10]:
```suggestion
   <version_command>python '$__tool_directory__/spyboat_cli.py' --version</version_command>
```

File: tools/spyboat/SpyBOAT.xml (hunk 2 above)
Comment (bernt-matthias) [inline tools/spyboat/SpyBOAT.xml:12]:
```suggestion
        python '$__tool_directory__/spyboat_cli.py'
```

File: tools/spyboat/SpyBOAT.xml (hunk 3 above)
Comment (bernt-matthias) [inline tools/spyboat/SpyBOAT.xml:21]: Can you use a consistent indentation of 4 spaces?

File: tools/spyboat/SpyBOAT.xml [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+        <param name="movie" type="data" format="tiff" label="Movie to analyze"
+               help="Select a movie to wavelet analyze"/>
+
+        <!--
+        The following Wavelet parameters must have the same numerical type as defined in the
+        argparse parser in spyboat_cli.py
+        -->
+
+        <param name="gauss_sigma" type="float" label="Sigma" 
+               help="Width of the Gaussian smoothing kernel, leave blank if no pre-smoothing desired." optional="true">
```
Comment (bernt-matthias) [inline tools/spyboat/SpyBOAT.xml:67]: Don't know if optional and a validator can be combined. Instead of the validator you may also use min and max attributes in the param itself.

File: tools/spyboat/SpyBOAT.xml [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+               help="Downsamples the input movie by this factor given in %, leave blank to disable downsampling"
+               optional="true">
+            <validator type="in_range" min="0" max="99"/>
+        </param>
+
+        <conditional name="masking">
+            <param label="Masking the output" name="selection_mode" type="select" help="Create a mask from the (blurred/rescaled) input to mask out regions without oscillations/signal in the output">
+                <option selected="true" value="no_masking">No masking</option>
+                <option value="create_static_mask">Create a static mask</option>
+		<option value="create_dynamic_mask">Create a dynamic mask</option>
```
Comment (bgruening) [inline tools/spyboat/SpyBOAT.xml:81]: indentation is here off too

File: tools/spyboat/SpyBOAT.xml [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+    <inputs>
+        <param name="movie" type="data" format="tiff" label="Movie to analyze"
+               help="Select a movie to wavelet analyze"/>
+
+        <!--
+        The following Wavelet parameters must have the same numerical type as defined in the
+        argparse parser in spyboat_cli.py
+        -->
+
+        <param name="gauss_sigma" type="float" label="Sigma"
```
Comment (bernt-matthias) [inline tools/spyboat/SpyBOAT.xml:66]:
```suggestion
        <param argument="gauss_sigma" type="float" label="Sigma" 
```

if no name is given it is autogenerated from argument (leading dashes are removed and internal dashes are replaced by underscore).

Comment (bgruening) [inline tools/spyboat/SpyBOAT.xml:66]: you can actually add min/max directly as an attribute to the param

File: tools/spyboat/SpyBOAT.xml [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+	  --html_fname $html_out
+	  --report_img_path '$html_out.extra_files_path'
+
+          --ncpu "\${GALAXY_SLOTS:-8}"
+          > $log
+
+	  ]]></command>
+    <!-- "\${GALAXY_SLOTS:-8}" -->
+    <inputs>
+        <param name="movie" type="data" format="tiff" label="Movie to analyze"
```
Comment (bernt-matthias) [inline tools/spyboat/SpyBOAT.xml:58]:
```suggestion
        <param name="movie" argument="--input_path" type="data" format="tiff" label="Movie to analyze"
```

File: tools/spyboat/SpyBOAT.xml [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+            <validator type="in_range" min="0" max="9999999"/>
+        </param>
+
+        <param name="win_size" type="float" value="" label="Sliding window size"
+               help="Window size for amplitude envelope estimation, leave blank to disable normalization "
+               optional="true">
+            <validator type="in_range" min="0" max="9999999"/>
+        </param>
+
+        <conditional name="save_preprocessed">
```
Comment (bgruening) [inline tools/spyboat/SpyBOAT.xml:131]: why is a conditional needed when you don't use the "when" cases?

File: tools/spyboat/SpyBOAT.xml [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
+
+        <param name="gauss_sigma" type="float" label="Sigma" 
+               help="Width of the Gaussian smoothing kernel, leave blank if no pre-smoothing desired." optional="true">
+            <validator type="in_range" min="0" max="9999999"/>
+        </param>
+
+        <param name="rescale_factor" type="integer" value="" label="Rescale factor"
+               help="Downsamples the input movie by this factor given in %, leave blank to disable downsampling"
+               optional="true">
+            <validator type="in_range" min="0" max="99"/>
```
Comment (bernt-matthias) [inline tools/spyboat/SpyBOAT.xml:74]: same as above: min and max ... optional?

File: tools/spyboat/SpyBOAT.xml [hunk 12]
```diff
@@ -0,0 +1,248 @@
...
+        </param>
+
+        <param name="win_size" type="float" value="" label="Sliding window size"
+               help="Window size for amplitude envelope estimation, leave blank to disable normalization "
+               optional="true">
+            <validator type="in_range" min="0" max="9999999"/>
+        </param>
+
+        <conditional name="save_preprocessed">
+            <param label="Save preprocessed input" name="selection" type="select" help="Save the blurred/rescaled input to history? ">
```
Comment (bgruening) [inline tools/spyboat/SpyBOAT.xml:132]: this could be a simple boolean, without conditional and without select.

File: tools/spyboat/SpyBOAT.xml [hunk 13]
```diff
@@ -0,0 +1,248 @@
...
+    <outputs>
+        <data name="phase_out" format="tiff" label="${movie.name[:-4]}_phase"/>
+        <data name="period_out" format="tiff" label="${movie.name[:-4]}_period"/>
+        <data name="power_out" format="tiff" label="${movie.name[:-4]}_power"/>
+        <data name="amplitude_out" format="tiff" label="${movie.name[:-4]}_amplitude"/>
+	<data name="preprocessed_out" format="tiff" label="${movie.namee[:-4]}_preproc">
+	  <filter>save_preprocessed['selection'] == 'Yes'</filter>
+	</data>
+	
+        <data name="log" format="txt" label="log ${movie.name[:-4]}.txt"/>
```
Comment (bgruening) [inline tools/spyboat/SpyBOAT.xml:153]: can we make more outputs optional, like this log file?

You could offer one bigger select param with multiple=true so the user can select multiple outputs

File: tools/spyboat/SpyBOAT.xml (hunk 10 above)
Comment (bernt-matthias) [inline tools/spyboat/SpyBOAT.xml:131]: Instead of the complete conditional you could simply have a single boolean

File: tools/spyboat/SpyBOAT.xml [hunk 14]
```diff
@@ -0,0 +1,248 @@
...
+
+    </outputs>
+
+    <tests>
+        <test>
+            <param name="movie" value="test-movie.tif" ftype="tiff"/>
+            <output name="phase_out" file="phase_out.tif" ftype="tiff" compare="sim_size"/>
+            <output name="period_out" file="period_out.tif" ftype="tiff" compare="sim_size"/>
+            <output name="power_out" file="power_out.tif" ftype="tiff" compare="sim_size"/>
+            <output name="amplitude_out" file="amplitude_out.tif" ftype="tiff" compare="sim_size"/>
```
Comment (bgruening) [inline tools/spyboat/SpyBOAT.xml:164]: you could use `has_size` here: https://docs.galaxyproject.org/en/latest/dev/schema.html#tool-tests-test-output-assert-contents

File: tools/spyboat/SpyBOAT.xml [hunk 15]
```diff
@@ -0,0 +1,248 @@
...
+	    <!-- there is an output filter picking up the choice -->
+	    <when value="No">
+	    </when>
+	    <when value="Yes">
+	    </when>	    
+	</conditional>
+    </inputs>
+
+    <outputs>
+        <data name="phase_out" format="tiff" label="${movie.name[:-4]}_phase"/>
```
Comment (bernt-matthias) [inline tools/spyboat/SpyBOAT.xml:145]: Please cange labels to something like `label="${tool.name} on ${on_string}: Phase"`

File: tools/spyboat/SpyBOAT.xml [hunk 16]
```diff
@@ -0,0 +1,248 @@
...
+        Length (in time units, e.g. hours) of the sliding window to estimate
+	an amplitude envelope. The signal then gets normalized by this envelope,
+        meaning that all amplitudes will be around ~1 after the transform. The 
+        advantage is that signal with strong amplitude trends will have more
+	meaningful Wavelet powers after normalization.
+
+	]]></help>
+    <citations>
+
+     <citation type="bibtex"> @article{monke2020optimal,
```
Comment (bgruening) [inline tools/spyboat/SpyBOAT.xml:239]: please use the DOI: 10.1101/2020.04.29.067744

File: tools/spyboat/SpyBOAT.xml [hunk 17]
```diff
@@ -0,0 +1,248 @@
...
+	  <filter>save_preprocessed['selection'] == 'Yes'</filter>
+	</data>
+	
+        <data name="log" format="txt" label="log ${movie.name[:-4]}.txt"/>
+	<data format="html" name="html_out" label="Report ${movie.name[:-4]}"/>
+
+    </outputs>
+
+    <tests>
+        <test>
```
Comment (bernt-matthias) [inline tools/spyboat/SpyBOAT.xml:159]: Add expect_num_outputs. Can you also add a test case where the html is written?

File: tools/spyboat/notes.md [hunk 18]
```diff
@@ -0,0 +1,27 @@
+## Notes for setting up the SpyBOAT Galaxy tool
+
+### Toolshed(s)
```
Comment (bgruening) [inline tools/spyboat/notes.md:3]: can you remove this file and replace it with a `.shed.yml` file. 

https://github.com/galaxyproject/tools-iuc/blob/master/tools/abricate/.shed.yml

We use those files to populate the toolshed metadata and only if you have such a file we run the tests automatically :)

File: tools/spyboat/notes.md [hunk 19]
```diff
@@ -0,0 +1,27 @@
+## Notes for setting up the SpyBOAT Galaxy tool
```
Comment (bernt-matthias) [inline tools/spyboat/notes.md:1]: I guess this file is not needed.

**Review by bgruening — COMMENTED** [review]
Thanks @tensionhead ... small review included

File: tools/spyboat/SpyBOAT.xml (hunk 16 above)
Comment (bernt-matthias) [inline tools/spyboat/SpyBOAT.xml:239]: If possible just use the doi approach.

File: tools/spyboat/output_report.py [hunk 20]
```diff
@@ -0,0 +1,250 @@
+''' Produces plots and a summary html 'headless' '''
```
Comment (bernt-matthias) [inline tools/spyboat/output_report.py:1]: Where is this file used?

File: tools/spyboat/OutputReport.html [hunk 21]
```diff
@@ -0,0 +1,271 @@
+
```
Comment (bernt-matthias) [inline tools/spyboat/OutputReport.html:1]: Is this file needed?

File: tools/spyboat/spyboat_cli.py [hunk 22]
```diff
@@ -0,0 +1,211 @@
...
+## Gets interfaced by Galaxy or can be used for bash scripting
+import argparse
+import sys, os
+import logging
+
+from skimage import io
+from numpy import float32
+
+import spyboat
+import output_report
```
Comment (bernt-matthias) [inline tools/spyboat/spyboat_cli.py:12]: Does this local import work?

File: tools/spyboat/styles.css [hunk 23]
```diff
@@ -0,0 +1,50 @@
+
```
Comment (bernt-matthias) [inline tools/spyboat/styles.css:1]: Not sure if it works this way. Maybe you need to put this in the extra files path of the generated html page?

File: tools/spyboat/output_report.py (hunk 20 above)
Comment (tensionhead) [reply tools/spyboat/output_report.py:1]: in the spyboat_cli.py, it generates the html and loads the figures from 
                                  `--report_img_path '$html_out.extra_files_path'`

File: tools/spyboat/OutputReport.html (hunk 21 above)
Comment (tensionhead) [reply tools/spyboat/OutputReport.html:1]: not really..

File: tools/spyboat/spyboat_cli.py (hunk 22 above)
Comment (tensionhead) [reply tools/spyboat/spyboat_cli.py:12]: yes, on my local galaxy instance that works just fine

File: tools/spyboat/styles.css (hunk 23 above)
Comment (tensionhead) [reply tools/spyboat/styles.css:1]: yeah this doesn't work, that's why the css is defined right in the huge htmk f-string of `create_html` inside `output_report.py`. I will remove it!

File: tools/spyboat/SpyBOAT.xml (hunk 3 above)
Comment (tensionhead) [reply tools/spyboat/SpyBOAT.xml:21]: yes, I will go through the whole xml again..

File: tools/spyboat/notes.md (hunk 18 above)
Comment (tensionhead) [reply tools/spyboat/notes.md:3]: ah yes, of course :)

File: tools/spyboat/SpyBOAT.xml (hunk 13 above)
Comment (tensionhead) [reply tools/spyboat/SpyBOAT.xml:153]: yeah that's probably a good idea in general, as some people might only be interested in say the period output. This will probably take me a little while..

File: tools/spyboat/SpyBOAT.xml (hunk 14 above)
Comment (tensionhead) [reply tools/spyboat/SpyBOAT.xml:164]: Ok I will look into it!

**Comment by tensionhead** [comment]
Pep8 collides with how matplotlib backends work: https://stackoverflow.com/questions/39305810/matplotlib-use-required-before-other-imports-clashes-with-pep8-ignore-or-fix I don't want to switch to the experimental feature 'switch-backend' so I suggest to ignore this?

**Comment by bgruening** [comment]
@tensionhead in very wired cases you can skip linting for single lines with https://stackoverflow.com/questions/18444840/how-to-disable-a-pep8-error-in-a-specific-file

File: tools/spyboat/.shed.yml [hunk 24]
```diff
@@ -0,0 +1,4 @@
+categories: [Imaging, Graphics]
+description: spyboat
+name: spyboat
+owner: gregor.m
```
Comment (bgruening) [inline tools/spyboat/.shed.yml:4]: This needs to be `iuc`, Can you please also add the other fields and a longer description. categories should be one of those: https://toolshed.g2.bx.psu.edu/

Comment (tensionhead) [reply tools/spyboat/.shed.yml:4]: Ok, the categories I already picked from that page. Can you elaborate on 'the other fields', thx!

Comment (tensionhead) [reply tools/spyboat/.shed.yml]: Ah I see, sorry I used `planemo shed_init` and somehow assumed all the required fields are there..

**Comment by bgruening** [comment]
@tensionhead let me know if this is ok for you. If this is merged you will find the next step here: https://galaxyproject.eu/posts/2020/08/22/three-steps-to-galaxify-your-tool/ :)

**Comment by tensionhead** [comment]
Looks good to me!

**Review by bgruening — APPROVED** [review]
